embedded-tests/src/test/java/org/apache/druid/testing/embedded/indexing/KafkaDataFormatsTest.java

```diff
@@ -0,0 +1,716 @@
+  private KafkaResource kafkaServer;
+  private SchemaRegistryResource schemaRegistry;
+
+  @BeforeAll
+  public static void beforeAll()
+  {
+    jsonMapper = TestHelper.JSON_MAPPER;
+  }
+
+  @Override
+  public EmbeddedDruidCluster createCluster()
+  {
+    final EmbeddedDruidCluster cluster = EmbeddedDruidCluster.withEmbeddedDerbyAndZookeeper().useContainerFriendlyHostname();
+
+    kafkaServer = new KafkaResource()
+    {
+      @Override
+      public void start()
+      {
+        super.start();
+      }
+
+      @Override
+      public void stop()
+      {
+        super.stop();
+      }
+    };
+
+    schemaRegistry = new SchemaRegistryResource()
+    {
+      @Override
+      public void start()
+      {
+        this.kafkaContainer = kafkaServer.getContainer();
+        super.start();
+      }
+
+      @Override
+      public void stop()
+      {
+        super.stop();
+      }
+    };
+
+    indexer.addProperty("druid.segment.handoff.pollDuration", "PT0.1s")
+           .addProperty("druid.worker.capacity", "10");
+    overlord.addProperty("druid.indexer.task.default.context", "{\"useConcurrentLocks\": true}")
+            .addProperty("druid.manager.segments.useIncrementalCache", "ifSynced")
+            .addProperty("druid.manager.segments.pollDuration", "PT0.1s")
+            .addProperty("druid.manager.segments.killUnused.enabled", "true")
+            .addProperty("druid.manager.segments.killUnused.bufferPeriod", "PT0.1s")
+            .addProperty("druid.manager.segments.killUnused.dutyPeriod", "PT1s");
+    coordinator.addProperty("druid.manager.segments.useIncrementalCache", "ifSynced");
+    cluster.addExtension(KafkaIndexTaskModule.class)
+           .addExtension(ProtobufExtensionsModule.class)
+           .addExtension(AvroExtensionsModule.class)
+           .useLatchableEmitter()
+           .addCommonProperty("druid.monitoring.emissionPeriod", "PT0.1s")
+           .addCommonProperty(
+               "druid.monitoring.monitors",
+               "[\"org.apache.druid.java.util.metrics.JvmMonitor\","
+               + "\"org.apache.druid.server.metrics.TaskCountStatsMonitor\"]"
+           )
+           .addResource(kafkaServer)
+           .addResource(schemaRegistry)
+           .addServer(coordinator)
+           .addServer(overlord)
+           .addServer(indexer)
+           .addServer(broker)
+           .addServer(historical);
+
+    return cluster;
+  }
+
+  @Test
+  @Timeout(value = 30, unit = TimeUnit.SECONDS)
+  public void test_indexKafka_avroDataFormat_withParser() throws IOException
+  {
+    new AvroExtensionsModule().getJacksonModules().forEach(jsonMapper::registerModule);
+    kafkaServer.createTopicWithPartitions(dataSource, 3);
+    EventSerializer serializer = jsonMapper.readValue("{\"type\": \"avro\"}", EventSerializer.class);
+    int recordCount = generateStreamAndPublishToKafka(dataSource, serializer, false);
+    String jsonString = "{"
```

Comment:
For all of the JSON strings, please try to use an existing builder or add a new one if necessary.
You may add it in `Resources` class or a new class which sits next to `TaskBuilder`.